### PR TITLE
mastodon,nixos/mastodon: init at 3.3.0

### PR DESCRIPTION
**Pull request by erictapen**
###### Motivation for this change

Quick discussion in the `#nix-community` IRC turned out that there is actually interest in merging https://github.com/NixOS/nixpkgs/pull/78810. @adisbladis offered to review this!

Also pinging @alyssais @Kiwi as they were also involved.

So I rebased on latest master, remove the tests and manual, as they are currently broken und added myself as maintainer.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112898` at 048c542f run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/917167b9738a0254a9078ffa39b5390d">mastodon</a></li></li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/servers/mastodon/default.nix:80:3:
```
   |
80 |   buildPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/servers/mastodon/default.nix:102:3:
```
    |
102 |   installPhase = ''
    |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`agpl3` is a deprecated license, check if project uses `agpl3Plus` or `agpl3Only` and change `meta.license` accordingly.

Near pkgs/servers/mastodon/default.nix:110:5:
```
    |
110 |     license = licenses.agpl3;
    |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `fetchFromGitHub`.
Near pkgs/servers/mastodon/default.nix:1:44:
```
  |
1 | { stdenv, lib, nodejs-slim, mkYarnPackage, fetchFromGitHub, bundlerEnv,
  |                                            ^
```
</li>
  </ul>
</details>

File: pkgs/servers/mastodon/default.nix [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  propagatedBuildInputs = [ imagemagick ffmpeg file mastodon-gems.wrappedRuby ];
+  installPhase = ''
+    mkdir -p $out
+    cp -r * $out/
+  '';
+
+  meta = with lib; {
+    description = "Self-hosted, globally interconnected microblogging software based on ActivityPub";
+    homepage = "https://joinmastodon.org";
+    license = licenses.agpl3;
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/default.nix:108]: agpl3Plus or agpl3Only?

File: pkgs/servers/mastodon/update.nix [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  '';
+  phases = [ "installPhase" ];
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ yarn2nix bundix coreutils diffutils nix-prefetch-github gnused jq ];
+
+  meta = {
+    maintainers = with stdenv.lib.maintainers; [ happy-river ];
+    description = "Utility to generate Nix expressions for Mastodon's dependencies";
+    platforms = stdenv.lib.platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/update.nix:19]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/servers/mastodon/update.nix [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    patchShebangs $out/bin/update.sh
+    wrapProgram $out/bin/update.sh --prefix PATH : ${lib.makeBinPath buildInputs}
+  '';
+  phases = [ "installPhase" ];
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ yarn2nix bundix coreutils diffutils nix-prefetch-github gnused jq ];
+
+  meta = {
+    maintainers = with stdenv.lib.maintainers; [ happy-river ];
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/update.nix:17]:
```suggestion
  meta = with lib; {
    maintainers = with maintainers; [ happy-river ];
```

stdenv.lib is deprecated and no longer allowed.

File: pkgs/servers/mastodon/update.nix [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  diffutils, nix-prefetch-github, gnused, jq }:
+stdenv.mkDerivation rec {
+  name = "mastodon-update-script";
+  installPhase = ''
+    mkdir -p $out/bin
+    cp ${./update.sh} $out/bin/update.sh
+    patchShebangs $out/bin/update.sh
+    wrapProgram $out/bin/update.sh --prefix PATH : ${lib.makeBinPath buildInputs}
+  '';
+  phases = [ "installPhase" ];
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/update.nix:11]:
```suggestion
```

Please do not overwrite phases which breaks hooks.

File: pkgs/servers/mastodon/update.nix [hunk 5]
```diff
@@ -0,0 +1,21 @@
+{ pkgs, stdenv, lib, makeWrapper, yarn2nix, bundix, coreutils,
+  diffutils, nix-prefetch-github, gnused, jq }:
+stdenv.mkDerivation rec {
+  name = "mastodon-update-script";
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/update.nix:4]: Why is this not a shell script with nix-shell?

File: pkgs/servers/mastodon/update.sh [hunk 6]
```diff
@@ -0,0 +1,121 @@
+#!/bin/sh
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/update.sh:1]:
```suggestion
#!/usr/bin/env bash
```

File: pkgs/servers/mastodon/update.nix (hunk 5 above)
Comment (adisbladis) [reply pkgs/servers/mastodon/update.nix:4]: Whatever.

File: pkgs/servers/mastodon/update.nix (hunk 4 above)
Comment (adisbladis) [reply pkgs/servers/mastodon/update.nix:11]: It doesn't matter, this is essentially a `runCommand`, it's just not using that abstraction.

File: pkgs/servers/mastodon/update.nix (hunk 5 above)
Comment (bqv) [reply pkgs/servers/mastodon/update.nix:4]: 😂

**Pull request merged**